datafusion-examples/examples/flight_sql_server.rs

```diff
@@ -0,0 +1,551 @@
+
+    Ok(())
+}
+
+pub struct FlightSqlServiceImpl {
+    contexts: Arc<DashMap<String, Arc<SessionContext>>>,
+    statements: Arc<DashMap<String, LogicalPlan>>,
+    results: Arc<DashMap<String, Vec<RecordBatch>>>,
+}
+
+impl FlightSqlServiceImpl {
+    fn get_ctx(&self, handle: &str) -> Result<Arc<SessionContext>, Status> {
+        if let Some(context) = self.contexts.get(handle) {
+            Ok(context.clone())
+        } else {
+            Err(Status::internal(format!(
+                "Context handle not found: {handle}"
+            )))?
+        }
+    }
+
+    fn get_plan(&self, handle: &str) -> Result<LogicalPlan, Status> {
+        if let Some(plan) = self.statements.get(handle) {
+            Ok(plan.clone())
+        } else {
+            Err(Status::internal(format!(
+                "Statement handle not found: {handle}"
+            )))?
+        }
+    }
+
+    fn get_result(&self, handle: &str) -> Result<Vec<RecordBatch>, Status> {
+        if let Some(result) = self.results.get(handle) {
+            Ok(result.clone())
+        } else {
+            Err(Status::internal(format!(
+                "Request handle not found: {handle}"
+            )))?
+        }
+    }
+
+    fn remove_plan(&self, handle: &str) -> Result<(), Status> {
+        self.statements.remove(&handle.to_string());
+        Ok(())
+    }
+
+    fn remove_result(&self, handle: &str) -> Result<(), Status> {
+        self.results.remove(&handle.to_string());
+        Ok(())
+    }
+}
+
+#[tonic::async_trait]
+impl FlightSqlService for FlightSqlServiceImpl {
+    type FlightService = FlightSqlServiceImpl;
+
+    async fn do_get_fallback(
+        &self,
+        _request: Request<Ticket>,
+        message: Any,
+    ) -> Result<Response<<Self as FlightService>::DoGetStream>, Status> {
+        if !message.is::<FetchResults>() {
+            Err(Status::unimplemented(format!(
+                "do_get: The defined request is invalid: {}",
+                message.type_url
+            )))?
+        }
+
+        let fr: FetchResults = message
+            .unpack()
+            .map_err(|e| Status::internal(format!("{e:?}")))?
+            .ok_or_else(|| Status::internal("Expected FetchResults but got None!"))?;
+
+        let handle = fr.handle;
+
+        info!("getting results for {handle}");
+        let result = self.get_result(&handle)?;
+        // if we get an empty result, create an empty schema
+        let (schema, batches) = match result.get(0) {
+            None => (Schema::empty(), vec![]),
+            Some(batch) => ((*batch.schema()).clone(), result.clone()),
+        };
+
+        let flight_data = batches_to_flight_data(schema, batches)
+            .map_err(|e| status!("Could not convert batches", e))?
+            .into_iter()
+            .map(Ok);
+
+        let stream: Pin<Box<dyn Stream<Item = Result<FlightData, Status>> + Send>> =
+            Box::pin(stream::iter(flight_data));
```

Comment:
I think you might be able to use the relatively newly added `FlightDataEncoderBuilder` here to create the FlightData:

https://docs.rs/arrow-flight/32.0.0/arrow_flight/encode/struct.FlightDataEncoderBuilder.html

Pretty much like

```rust
// Get an input stream of Result<RecordBatch, FlightError>
let input_stream = futures::stream::iter(batches);

// Build a stream of `Result<FlightData>` (e.g. to return for do_get)
let flight_data_stream = FlightDataEncoderBuilder::new()
  .with_schema(shcema)
 .build(input_stream);

tonic::Response::new(flight_data_stream);
```

Reply:
I wasn't able to figure out how to use this :/ -- the result of FlightDataEncoderBuilder.build() is not accepted by tonic::Response::new(). My experience with futures and streams is pretty limited, so maybe I'm missing something.

Reply:
I'll give it a shot later today -- thanks @kmitchener

Reply:
I hit something like this recently. The difficult part is more the deeply nested type signatures than the async part IMO. I found some similar code and copy & paste solved it:

```
        let output = futures::stream::iter(vec![Ok(PutResult {
            app_metadata: Default::default(),
        })]);
        return Ok(Response::new(Box::pin(output)));
```

The type signature for that function turns into something like:

```
Result<Response<Pin<Box<dyn Stream<Item = Result<PutResult, Status>> + Send + 'static>>>, Status>
```

So a result of a pinned box holding an implementation of a Stream of Results of either PutResults or Statuses. Not exactly straight forward, I think it took me an hour to figure it out in the original code.

Reply:
@kmitchener  here is a PR to this branch https://github.com/kmitchener/arrow-datafusion/pull/1  that uses FlightDataEncder (I agree the stream nonsense is quite tricky when it comes to understanding what the compiler is telling you)

Reply:
that works! and I think the other implementation was buggy as well, as it would sometimes not return expected rows. I was in the midst of troubleshooting that, but your changes here resolved it.